File: robot_ros/hardware_interface/arduino_hardware_interface.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
import serial
import threading
import time
import re
from hardware_interface import SystemInterface, HardwareInfo, CallbackReturn
from rclpy.duration import Duration


class ArduinoHardwareInterface(SystemInterface):
    """
    Hardware interface per comunicare con Arduino utilizzando il protocollo
    del ros_arduino_bridge firmware.
    """
    
    def __init__(self):
        super().__init__()
        self.arduino = None
        self.left_wheel_pos = 0.0
        self.right_wheel_pos = 0.0
        self.left_wheel_vel = 0.0
        self.right_wheel_vel = 0.0
        self.left_wheel_cmd = 0.0
        self.right_wheel_cmd = 0.0
        
        # Parametri del robot
        self.wheel_radius = 0.065  # m
        self.wheel_separation = 0.17  # m
        self.ticks_per_rev = 3436
        
        # Comunicazione seriale
        self.port = "/dev/ttyUSB0"
        self.baudrate = 57600
        self.timeout = 1.0
        
        # Thread per lettura encoder
        self.reading_thread = None
        self.is_running = False
        
        # Buffer per i comandi
        self.command_lock = threading.Lock()
        
    def on_init(self, info: HardwareInfo) -> CallbackReturn:
        """Inizializzazione dell'interfaccia hardware"""
        
        # Leggi i parametri dal file URDF
        try:
            if "device" in info.hardware_parameters:
                self.port = info.hardware_parameters["device"]
            if "baud_rate" in info.hardware_parameters:
                self.baudrate = int(info.hardware_parameters["baud_rate"])
            if "timeout" in info.hardware_parameters:
                self.timeout = float(info.hardware_parameters["timeout"]) / 1000.0
            if "enc_counts_per_rev" in info.hardware_parameters:
                self.ticks_per_rev = int(info.hardware_parameters["enc_counts_per_rev"])
                
        except Exception as e:
            self.get_logger().error(f"Error reading parameters: {e}")
            return CallbackReturn.ERROR
            
        # Verifica che i joint siano corretti
        if len(info.joints) != 2:
            self.get_logger().error("Expected exactly 2 joints")
            return CallbackReturn.ERROR
            
        self.left_joint_name = info.joints[0].name
        self.right_joint_name = info.joints[1].name
        
        self.get_logger().info(f"Arduino Hardware Interface initialized")
        self.get_logger().info(f"Port: {self.port}, Baudrate: {self.baudrate}")
        self.get_logger().info(f"Joints: {self.left_joint_name}, {self.right_joint_name}")
        
        return CallbackReturn.SUCCESS
        
    def on_configure(self, previous_state) -> CallbackReturn:
        """Configurazione dell'interfaccia"""
        
        try:
            # Connetti ad Arduino
            self.arduino = serial.Serial(
                port=self.port,
                baudrate=self.baudrate,
                timeout=self.timeout
            )
            
            # Aspetta che Arduino si inizializzi
            time.sleep(2)
            
            # Pulisci il buffer
            self.arduino.flushInput()
            self.arduino.flushOutput()
            
            # Testa la connessione
            self.send_command("v")  # Richiedi versione
            time.sleep(0.1)
            
            # Leggi la risposta
            response = self.read_response()
            if response:
                self.get_logger().info(f"Arduino connected: {response}")
            else:
                self.get_logger().warn("No response from Arduino, continuing anyway...")
                
            # Reset encoders
            self.send_command("r")
            
            self.get_logger().info("Arduino configured successfully")
            return CallbackReturn.SUCCESS
            
        except Exception as e:
            self.get_logger().error(f"Failed to configure Arduino: {e}")
            return CallbackReturn.ERROR
            
    def on_activate(self, previous_state) -> CallbackReturn:
        """Attivazione dell'interfaccia"""
        
        try:
            # Avvia il thread di lettura
            self.is_running = True
            self.reading_thread = threading.Thread(target=self.read_encoders_loop)
            self.reading_thread.daemon = True
            self.reading_thread.start()
            
            self.get_logger().info("Arduino Hardware Interface activated")
            return CallbackReturn.SUCCESS
            
        except Exception as e:
            self.get_logger().error(f"Failed to activate Arduino interface: {e}")
            return CallbackReturn.ERROR
            
    def on_deactivate(self, previous_state) -> CallbackReturn:
        """Disattivazione dell'interfaccia"""
        
        # Ferma i motori
        self.send_motor_commands(0, 0)
        
        # Ferma il thread di lettura
        self.is_running = False
        if self.reading_thread and self.reading_thread.is_alive():
            self.reading_thread.join(timeout=1.0)
            
        self.get_logger().info("Arduino Hardware Interface deactivated")
        return CallbackReturn.SUCCESS
        
    def on_cleanup(self, previous_state) -> CallbackReturn:
        """Pulizia dell'interfaccia"""
        
        if self.arduino and self.arduino.is_open:
            self.arduino.close()
            
        self.get_logger().info("Arduino Hardware Interface cleaned up")
        return CallbackReturn.SUCCESS
        
    def on_error(self, previous_state) -> CallbackReturn:
        """Gestione errori"""
        
        self.get_logger().error("Arduino Hardware Interface in error state")
        return CallbackReturn.SUCCESS
        
    def read(self, time, period) -> bool:
        """Lettura dello stato del hardware"""
        # I dati vengono aggiornati dal thread di lettura
        return True
        
    def write(self, time, period) -> bool:
        """Scrittura dei comandi al hardware"""
        
        try:
            with self.command_lock:
                # Converti velocità angolare delle ruote in PWM
                left_pwm = self.velocity_to_pwm(self.left_wheel_cmd)
                right_pwm = self.velocity_to_pwm(self.right_wheel_cmd)
                
                # Invia comandi ai motori
                self.send_motor_commands(left_pwm, right_pwm)
                
            return True
            
        except Exception as e:
            self.get_logger().error(f"Error in write: {e}")
            return False
            
    def send_command(self, command):
        """Invia un comando ad Arduino"""
        if self.arduino and self.arduino.is_open:
            try:
                self.arduino.write((command + '\r').encode())
                self.arduino.flush()
            except Exception as e:
                self.get_logger().error(f"Error sending command '{command}': {e}")
                
    def read_response(self):
        """Leggi una risposta da Arduino"""
        if self.arduino and self.arduino.is_open:
            try:
                response = self.arduino.readline().decode().strip()
                return response
            except Exception as e:
                self.get_logger().error(f"Error reading response: {e}")
        return None
        
    def send_motor_commands(self, left_pwm, right_pwm):
        """Invia comandi PWM ai motori"""
        # Limita i valori PWM
        left_pwm = max(-255, min(255, int(left_pwm)))
        right_pwm = max(-255, min(255, int(right_pwm)))
        
        # Formato comando: m left_pwm right_pwm
        command = f"m {left_pwm} {right_pwm}"
        self.send_command(command)
        
    def read_encoders_loop(self):
        """Loop per leggere gli encoder in un thread separato"""
        
        while self.is_running:
            try:
                if self.arduino and self.arduino.is_open:
                    # Richiedi lettura encoder
                    self.send_command("e")
                    
                    # Leggi risposta
                    response = self.read_response()
                    
                    if response:
                        # Parsa la risposta degli encoder
                        # Formato atteso: "left_ticks right_ticks"
                        parts = response.split()
                        if len(parts) >= 2:
                            try:
                                left_ticks = int(parts[0])
                                right_ticks = int(parts[1])
                                
                                # Converti ticks in posizione (radianti)
                                left_pos = (left_ticks / self.ticks_per_rev) * 2 * 3.14159
                                right_pos = (right_ticks / self.ticks_per_rev) * 2 * 3.14159
                                
                                # Calcola velocità (differenza dalla lettura precedente)
                                dt = 0.1  # 10 Hz
                                left_vel = (left_pos - self.left_wheel_pos) / dt
                                right_vel = (right_pos - self.right_wheel_pos) / dt
                                
                                # Aggiorna i valori
                                self.left_wheel_pos = left_pos
                                self.right_wheel_pos = right_pos
                                self.left_wheel_vel = left_vel
                                self.right_wheel_vel = right_vel
                                
                            except ValueError:
                                self.get_logger().warn(f"Invalid encoder response: {response}")
                                
                time.sleep(0.1)  # 10 Hz
                
            except Exception as e:
                self.get_logger().error(f"Error in encoder reading loop: {e}")
                time.sleep(1.0)
                
    def velocity_to_pwm(self, velocity):
        """Converti velocità angolare (rad/s) in PWM (-255 a 255)"""
        # Fattore di conversione empirico
        # Potrebbe essere necessario calibrarlo per il tuo robot
        max_velocity = 10.0  # rad/s
        pwm = (velocity / max_velocity) * 255
        return max(-255, min(255, pwm))
        
    def export_state_interfaces(self):
        """Esporta le interfacce di stato"""
        state_interfaces = []
        
        # Interfacce per la ruota sinistra
        state_interfaces.append(
            HardwareInterface(
                name=f"{self.left_joint_name}/position",
                initial_value=0.0
            )
        )
        state_interfaces.append(
            HardwareInterface(
                name=f"{self.left_joint_name}/velocity", 
                initial_value=0.0
            )
        )
        
        # Interfacce per la ruota destra
        state_interfaces.append(
            HardwareInterface(
                name=f"{self.right_joint_name}/position",
                initial_value=0.0
            )
        )
        state_interfaces.append(
            HardwareInterface(
                name=f"{self.right_joint_name}/velocity",
                initial_value=0.0
            )
        )
        
        return state_interfaces
        
    def export_command_interfaces(self):
        """Esporta le interfacce di comando"""
        command_interfaces = []
        
        # Interfacce per la ruota sinistra
        command_interfaces.append(
            HardwareInterface(
                name=f"{self.left_joint_name}/velocity",
                initial_value=0.0
            )
        )
        
        # Interfacce per la ruota destra  
        command_interfaces.append(
            HardwareInterface(
                name=f"{self.right_joint_name}/velocity",
                initial_value=0.0
            )
        )
        
        return command_interfaces
        
    def get_state_interface_value(self, interface_name):
        """Ottieni il valore di un'interfaccia di stato"""
        if "left_wheel_joint/position" in interface_name:
            return self.left_wheel_pos
        elif "left_wheel_joint/velocity" in interface_name:
            return self.left_wheel_vel
        elif "right_wheel_joint/position" in interface_name:
            return self.right_wheel_pos
        elif "right_wheel_joint/velocity" in interface_name:
            return self.right_wheel_vel
        return 0.0
        
    def set_command_interface_value(self, interface_name, value):
        """Imposta il valore di un'interfaccia di comando"""
        if "left_wheel_joint/velocity" in interface_name:
            self.left_wheel_cmd = value
        elif "right_wheel_joint/velocity" in interface_name:
            self.right_wheel_cmd = value

File: robot_ros/hardware_interface/arduino_hardware_interface.cpp
#include "arduino_hardware_interface.hpp"

#include <chrono>
#include <cstring>
#include <sstream>
#include <thread>
#include <errno.h>

namespace arduino_hardware_interface
{

hardware_interface::CallbackReturn ArduinoHardwareInterface::on_init(
  const hardware_interface::HardwareInfo & info)
{
  if (
    hardware_interface::SystemInterface::on_init(info) !=
    hardware_interface::CallbackReturn::SUCCESS)
  {
    return hardware_interface::CallbackReturn::ERROR;
  }

  // Leggi parametri
  device_ = info_.hardware_parameters.at("device");
  baud_rate_ = std::stoi(info_.hardware_parameters.at("baud_rate"));
  timeout_ = std::stod(info_.hardware_parameters.at("timeout")) / 1000.0;
  
  // Parametri opzionali
  wheel_radius_ = 0.065;  // 65mm default
  wheel_separation_ = 0.17;  // 170mm default

  // Verifica che ci siano esattamente 2 joint
  if (info_.joints.size() != 2) {
    RCLCPP_ERROR(logger_, "Expected exactly 2 joints, got %zu", info_.joints.size());
    return hardware_interface::CallbackReturn::ERROR;
  }

  // Ottieni i nomi dei joint
  left_wheel_name_ = info_.joints[0].name;
  right_wheel_name_ = info_.joints[1].name;

  // Inizializza i vettori di stato
  hw_positions_.resize(info_.joints.size(), 0.0);
  hw_velocities_.resize(info_.joints.size(), 0.0);
  hw_commands_.resize(info_.joints.size(), 0.0);

  RCLCPP_INFO(logger_, "Arduino Hardware Interface initialized");
  RCLCPP_INFO(logger_, "Device: %s, Baud: %d", device_.c_str(), baud_rate_);
  RCLCPP_INFO(logger_, "Left wheel: %s, Right wheel: %s", 
              left_wheel_name_.c_str(), right_wheel_name_.c_str());

  return hardware_interface::CallbackReturn::SUCCESS;
}

std::vector<hardware_interface::StateInterface> ArduinoHardwareInterface::export_state_interfaces()
{
  std::vector<hardware_interface::StateInterface> state_interfaces;
  
  for (size_t i = 0; i < info_.joints.size(); i++) {
    state_interfaces.emplace_back(hardware_interface::StateInterface(
      info_.joints[i].name, "position", &hw_positions_[i]));
    state_interfaces.emplace_back(hardware_interface::StateInterface(
      info_.joints[i].name, "velocity", &hw_velocities_[i]));
  }

  return state_interfaces;
}

std::vector<hardware_interface::CommandInterface> ArduinoHardwareInterface::export_command_interfaces()
{
  std::vector<hardware_interface::CommandInterface> command_interfaces;
  
  for (size_t i = 0; i < info_.joints.size(); i++) {
    command_interfaces.emplace_back(hardware_interface::CommandInterface(
      info_.joints[i].name, "velocity", &hw_commands_[i]));
  }

  return command_interfaces;
}

hardware_interface::CallbackReturn ArduinoHardwareInterface::on_activate(
  const rclcpp_lifecycle::State & /*previous_state*/)
{
  RCLCPP_INFO(logger_, "Activating Arduino Hardware Interface...");
  
  if (!connect_arduino()) {
    return hardware_interface::CallbackReturn::ERROR;
  }

  // Test connessione Arduino con comando baudrate
  send_command("b");
  std::this_thread::sleep_for(std::chrono::milliseconds(500));
  
  // Reset encoders se presenti
  send_command("r");
  std::this_thread::sleep_for(std::chrono::milliseconds(200));

  RCLCPP_INFO(logger_, "Arduino Hardware Interface activated");
  return hardware_interface::CallbackReturn::SUCCESS;
}

hardware_interface::CallbackReturn ArduinoHardwareInterface::on_deactivate(
  const rclcpp_lifecycle::State & /*previous_state*/)
{
  RCLCPP_INFO(logger_, "Deactivating Arduino Hardware Interface...");

  // Ferma i motori con comando PWM
  send_motor_commands(0, 0);
  std::this_thread::sleep_for(std::chrono::milliseconds(200));

  disconnect_arduino();

  RCLCPP_INFO(logger_, "Arduino Hardware Interface deactivated");
  return hardware_interface::CallbackReturn::SUCCESS;
}

hardware_interface::return_type ArduinoHardwareInterface::read(
  const rclcpp::Time & /*time*/, const rclcpp::Duration & period)
{
  // Simula movimento delle ruote in base ai comandi
  // In un sistema reale qui leggeresti gli encoder
  double dt = period.seconds();
  
  if (dt > 0) {
    // Simula il movimento basato sui comandi attuali
    for (size_t i = 0; i < hw_commands_.size(); i++) {
      hw_velocities_[i] = hw_commands_[i]; // Velocità = comando
      hw_positions_[i] += hw_velocities_[i] * dt; // Integra posizione
    }
  }

  return hardware_interface::return_type::OK;
}

hardware_interface::return_type ArduinoHardwareInterface::write(
  const rclcpp::Time & /*time*/, const rclcpp::Duration & /*period*/)
{
  // DEBUG: Logga velocità ricevute dai controller
  if (hw_commands_[0] != 0.0 || hw_commands_[1] != 0.0) {
    RCLCPP_INFO(logger_, "RECEIVED VELOCITIES: Left=%.3f rad/s, Right=%.3f rad/s", 
                hw_commands_[0], hw_commands_[1]);
  }
  
  // Converti velocità in PWM e invia comandi
  int left_pwm = velocity_to_pwm(hw_commands_[0]);
  int right_pwm = velocity_to_pwm(hw_commands_[1]);
  
  send_motor_commands(left_pwm, right_pwm);

  return hardware_interface::return_type::OK;
}

bool ArduinoHardwareInterface::connect_arduino()
{
  // Apri porta seriale con flag diversi per evitare lock
  serial_fd_ = open(device_.c_str(), O_RDWR | O_NOCTTY);
  
  if (serial_fd_ == -1) {
    RCLCPP_ERROR(logger_, "Failed to open serial port: %s", device_.c_str());
    return false;
  }

  // Configura porta seriale in modo più robusto
  struct termios options;
  memset(&options, 0, sizeof(options));
  
  // Imposta baudrate
  speed_t speed;
  switch (baud_rate_) {
    case 9600:   speed = B9600;   break;
    case 19200:  speed = B19200;  break;
    case 38400:  speed = B38400;  break;
    case 57600:  speed = B57600;  break;
    case 115200: speed = B115200; break;
    default:
      RCLCPP_ERROR(logger_, "Unsupported baud rate: %d", baud_rate_);
      close(serial_fd_);
      return false;
  }
  
  cfsetispeed(&options, speed);
  cfsetospeed(&options, speed);
  
  // Configura parametri seriali - configurazione robusta
  options.c_cflag = CS8 | CLOCAL | CREAD;  // 8 bit, locale, enable receiver
  options.c_iflag = IGNPAR;                // Ignore parity errors
  options.c_oflag = 0;                     // Raw output
  options.c_lflag = 0;                     // Raw input
  
  // Timeout
  options.c_cc[VMIN] = 0;  // Non-blocking read
  options.c_cc[VTIME] = 1; // 0.1 second timeout

  // Applica configurazione immediatamente
  if (tcsetattr(serial_fd_, TCSANOW, &options) != 0) {
    RCLCPP_ERROR(logger_, "Failed to configure serial port");
    close(serial_fd_);
    return false;
  }
  
  // Aspetta che Arduino si stabilizzi
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  
  // Pulisci buffer entrambe le direzioni
  tcflush(serial_fd_, TCIOFLUSH);
  
  // Test di connessione con comando semplice
  std::string test_cmd = "b\n";
  if (::write(serial_fd_, test_cmd.c_str(), test_cmd.length()) == -1) {
    RCLCPP_ERROR(logger_, "Failed to test Arduino connection");
    close(serial_fd_);
    return false;
  }

  RCLCPP_INFO(logger_, "Arduino connected successfully on %s at %d baud", device_.c_str(), baud_rate_);
  return true;
}

void ArduinoHardwareInterface::disconnect_arduino()
{
  if (serial_fd_ != -1) {
    close(serial_fd_);
    serial_fd_ = -1;
  }
}

bool ArduinoHardwareInterface::send_command(const std::string& command)
{
  if (serial_fd_ == -1) return false;
  
  std::string cmd = command + "\n";
  size_t total_bytes = cmd.length();
  size_t bytes_sent = 0;
  int attempts = 0;
  const int max_attempts = 3;
  
  // Prova a inviare con retry per gestire "Partial write"
  while (bytes_sent < total_bytes && attempts < max_attempts) {
    ssize_t bytes_written = ::write(serial_fd_, cmd.c_str() + bytes_sent, total_bytes - bytes_sent);
    
    if (bytes_written < 0) {
      if (errno == EAGAIN || errno == EWOULDBLOCK) {
        // Buffer pieno, aspetta un po'
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
        attempts++;
        continue;
      } else {
        RCLCPP_ERROR(logger_, "Write error for command: %s, errno: %d", command.c_str(), errno);
        return false;
      }
    }
    
    bytes_sent += bytes_written;
    attempts++;
    
    // Se non abbiamo inviato tutto, aspetta un po'
    if (bytes_sent < total_bytes) {
      std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
  }
  
  if (bytes_sent < total_bytes) {
    RCLCPP_WARN(logger_, "Partial write for command: %s (%zu/%zu bytes)", command.c_str(), bytes_sent, total_bytes);
    return false;
  }
  
  // Forza l'invio immediato dei dati seriali (sync invece di flush)
  fsync(serial_fd_);
  
  return true;
}

void ArduinoHardwareInterface::send_motor_commands(int left_pwm, int right_pwm)
{
  // Limita valori PWM
  left_pwm = std::max(-255, std::min(255, left_pwm));
  right_pwm = std::max(-255, std::min(255, right_pwm));
  
  RCLCPP_INFO(logger_, "Sending motor commands: LEFT=%d, RIGHT=%d", left_pwm, right_pwm);
  
  // PROTOCOLLO CORRETTO per ROSArduinoBridge!
  // Comando: "o left_pwm right_pwm" per controllo PWM diretto
  std::string motor_cmd = "o " + std::to_string(left_pwm) + " " + std::to_string(right_pwm);
  
  if (send_command(motor_cmd)) {
    RCLCPP_INFO(logger_, "Motor PWM command sent: %s", motor_cmd.c_str());
  } else {
    RCLCPP_WARN(logger_, "Failed to send motor command: %s", motor_cmd.c_str());
  }
}

int ArduinoHardwareInterface::velocity_to_pwm(double velocity)
{
  // Converti velocità angolare rad/s in PWM
  const double max_velocity = 5.0; // rad/s massima
  double pwm = (velocity / max_velocity) * 255.0;
  return static_cast<int>(std::max(-255.0, std::min(255.0, pwm)));
}

}  // namespace arduino_hardware_interface

// Macro per registrare il plugin
#include "pluginlib/class_list_macros.hpp"

PLUGINLIB_EXPORT_CLASS(
  arduino_hardware_interface::ArduinoHardwareInterface,
  hardware_interface::SystemInterface
)